src/NHibernate/Transaction/AdoNetWithSystemTransactionFactory.cs

```diff
@@ -0,0 +1,545 @@
+			foreach (var dependentSession in originatingSession.ConnectionManager.DependentSessions)
+			{
+				EnlistDependentSession(dependentSession, transactionContext);
+			}
+		}
+
+		/// <summary>
+		/// Create a transaction context for enlisting a session with a <see cref="System.Transactions.Transaction"/>,
+		/// and enlist the context in the transaction.
+		/// </summary>
+		/// <param name="session">The session to be enlisted.</param>
+		/// <param name="transaction">The transaction into which the context has to be enlisted.</param>
+		/// <returns>The created transaction context.</returns>
+		protected virtual ITransactionContext CreateAndEnlistMainContext(
+			ISessionImplementor session,
+			System.Transactions.Transaction transaction)
+		{
+			var transactionContext = new SystemTransactionContext(
+				session, transaction, SystemTransactionCompletionLockTimeout,
+				UseConnectionOnSystemTransactionPrepare);
+			transactionContext.AmbientTransaction.EnlistVolatile(
+				transactionContext,
+				UseConnectionOnSystemTransactionPrepare
+					? EnlistmentOptions.EnlistDuringPrepareRequired
+					: EnlistmentOptions.None);
+			return transactionContext;
+		}
+
+		private void EnlistDependentSession(ISessionImplementor dependentSession, ITransactionContext mainContext)
+		{
+			dependentSession.TransactionContext = CreateDependentContext(dependentSession, mainContext);
+			dependentSession.AfterTransactionBegin(null);
+		}
+
+		/// <summary>
+		/// Create a transaction context for a dependent session.
+		/// </summary>
+		/// <param name="dependentSession">The dependent session.</param>
+		/// <param name="mainContext">The context of the session owning the <see cref="ConnectionManager"/>.</param>
+		/// <returns>A dependent context for the session.</returns>
+		protected virtual ITransactionContext CreateDependentContext(ISessionImplementor dependentSession, ITransactionContext mainContext)
+		{
+			return new DependentContext(mainContext);
+		}
+
+		/// <inheritdoc />
+		public override bool IsInActiveSystemTransaction(ISessionImplementor session)
+			=> session.TransactionContext?.IsInActiveTransaction ?? false;
+
+		/// <inheritdoc />
+		public override void ExecuteWorkInIsolation(ISessionImplementor session, IIsolatedWork work, bool transacted)
+		{
+			using (var tx = new TransactionScope(TransactionScopeOption.Suppress))
+			{
+				base.ExecuteWorkInIsolation(session, work, transacted);
+				tx.Complete();
+			}
+		}
+
+		/// <summary>
+		/// Transaction context for enlisting a session with a system <see cref="System.Transactions.Transaction"/>.
+		/// It is meant for being the concrete class enlisted in the transaction.
+		/// </summary>
+		public class SystemTransactionContext : ITransactionContext, IEnlistmentNotification
+		{
+			/// <summary>
+			/// The transaction in which this context is enlisted.
+			/// </summary>
+			protected internal System.Transactions.Transaction AmbientTransaction { get; }
```

Comment:
Why not. My main concern is more that `originalTransaction` I have added. Functionally, that is the same, but technically, there is one (currently named `AmbientTransaction`) we have more control on the life-cycle (does not seem to be disposed of before we choose) while the other is more reliable for checking the transaction status (the `original` one).

The enlistment could likely be done with any of them. That is more the prepare scope which should be done with the clone. But I have not sufficiently experiment with that. (I have just tried no more using a clone at all at the beginning, and that was causing more issues too.)